here-naksha-lib-view/src/main/java/com/here/naksha/lib/view/ViewReadSession.java

```diff
@@ -0,0 +1,247 @@
+import com.here.naksha.lib.core.models.storage.HeapCacheCursor;
+import com.here.naksha.lib.core.models.storage.Notification;
+import com.here.naksha.lib.core.models.storage.ReadRequest;
+import com.here.naksha.lib.core.models.storage.Result;
+import com.here.naksha.lib.core.models.storage.XyzFeatureCodecFactory;
+import com.here.naksha.lib.core.storage.IReadSession;
+import com.here.naksha.lib.core.storage.ISession;
+import com.here.naksha.lib.view.concurrent.LayerReadRequest;
+import com.here.naksha.lib.view.concurrent.ParallelQueryExecutor;
+import com.here.naksha.lib.view.merge.MergeByStoragePriority;
+import com.here.naksha.lib.view.missing.ObligatoryLayersResolver;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.tuple.Pair;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * {@link  ViewReadSession} operates on {@link View}, it queries simultaneously all the storages.
+ * Then it tries to feeth missing features {@link MissingIdResolver} if needed.
+ * At the end {@link MergeOperation} is executed and single result returned.
+ * You can provide your own merge operation. The default is "take result from storage on the top". <br>
+ *
+ * <strong>Important:</strong> {@link ViewReadSession} will always return mutable cursor, this is the only way we can
+ * merge results from different storages and fetch missing by ids. Consider this example:
+ * Result from Storage A: [F_1, F_2, F_3, F_4]
+ * Result from Storage B: [F_2, F_4]
+ * Result from Storage C: [F_3, F_5]
+ * In this situation using Forward cursor would lead to N+1 issue, as after reading 1st row from each result we'd have
+ * to fetch missing F_1 from B and C.
+ * To be able to create query that fetches multiple missing features we have to know them first (by caching ahead of time) <br>
+ * <p>
+ * It might happen that feature has been moved (it's geometry changed). In such case after getting results for bbox
+ * query we have to query again for all features (by id) that was missing in a least one storage  result.
+ */
+public class ViewReadSession implements IReadSession {
+
+  private final View viewRef;
+
+  protected ParallelQueryExecutor parallelQueryExecutor;
+
+  protected Map<ViewLayer, IReadSession> subSessions;
+
+  protected ViewReadSession(@NotNull View viewRef, @Nullable NakshaContext context, boolean useMaster) {
+    this.viewRef = viewRef;
+    this.subSessions = new LinkedHashMap<>();
+    for (ViewLayer layer : viewRef.getViewCollection().getLayers()) {
+      subSessions.put(layer, layer.getStorage().newReadSession(context, useMaster));
+    }
+    this.parallelQueryExecutor = new ParallelQueryExecutor(viewRef);
+  }
+
+  @Override
+  public @NotNull Result execute(@NotNull ReadRequest<?> readRequest) {
```

Comment:
As a default implementation, I am of an opinion that this should apply default optimized use of resolver i.e. use IgnoreMissingResolver or OligatoryLayersResolver depending on whether it is GetById(s) request or not.